### ISSUE TITLE
react-boilerplate integration

### ISSUE THREAD

**Issue by thecodingwhale** [rt2zz/redux-persist#581]
I'm currently using https://github.com/react-boilerplate/react-boilerplate and I'm just wonder if anyone integrate redux-persist with it?

**Comment by rt2zz** [comment]
It should integrate easily per the instructions in the readme

**Comment by mouthzipper** [comment]
@thecodingwhale  any luck for the integration? Integration for v4.8.2 is straightforward but can't get it working for the v5.

**Comment by mouthzipper** [comment]
Integration with react-boilerplate:
1: edit reducers.js something like this:
`/**
 * Combine all reducers in this file and export the combined reducers.
 */
import { LOCATION_CHANGE } from 'react-router-redux';
import { persistCombineReducers } from 'redux-persist';
import session from 'redux-persist/lib/storage/session';

import globalReducer from 'containers/App/reducer';
import languageProviderReducer from 'containers/LanguageProvider/reducer';
import currencyToggleReducer from 'containers/CurrencyToggle/reducer';
import merge from 'lodash/merge';

/*
 * routeReducer
 *
 * The reducer merges route location changes into our immutable state.
 * The change is necessitated by moving to react-router-redux@5
 *
 */

// Initial routing state
const routeInitialState = {
  location: null,
};

const config = {
  key: 'root',
  storage: session,
};

function routeReducer(state = routeInitialState, action) {
  switch (action.type) {
    /* istanbul ignore next */
    case LOCATION_CHANGE:
      return merge({}, state, {
        location: action.payload,
      });
    default:
      return state;
  }
}

export default function createReducer(injectedReducers) {
  return persistCombineReducers(config, {
    route: routeReducer,
    global: globalReducer,
    language: languageProviderReducer,
    ...injectedReducers,
  });
}
`

2: edit configureStore.js something like this:
`/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware, compose } from 'redux';
import { routerMiddleware } from 'react-router-redux';
import createSagaMiddleware from 'redux-saga';
import { persistStore } from 'redux-persist';
import createReducer from './reducers';

const sagaMiddleware = createSagaMiddleware();

export default function configureStore(initialState = {}, history) {
  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [
    sagaMiddleware,
    routerMiddleware(history),
  ];

  const enhancers = [
    applyMiddleware(...middlewares),
  ];

  // If Redux DevTools Extension is installed use it, otherwise use Redux compose
  /* eslint-disable no-underscore-dangle */
  const composeEnhancers =
    process.env.NODE_ENV !== 'production' &&
    typeof window === 'object' &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
        // TODO Try to remove when `react-router-redux` is out of beta, LOCATION_CHANGE should not be fired more than once after hot reloading
        // Prevent recomputing reducers for `replaceReducer`
        shouldHotReload: false,
      })
      : compose;
  /* eslint-enable */

  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers)
  );

  // Extensions
  store.runSaga = sagaMiddleware.run;
  store.injectedReducers = {}; // Reducer registry
  store.injectedSagas = {}; // Saga registry

  // Make reducers hot reloadable, see http://mxs.is/googmo
  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      store.replaceReducer(createReducer(store.injectedReducers));
    });
  }
  const persistor = persistStore(store);
  return { persistor, store };
}
`
3: edit app.js
`import { PersistGate } from 'redux-persist/es/integration/react';

const { persistor, store } = configureStore(initialState, history);

const render = (messages) => {
  ReactDOM.render(
    <Provider store={store}>
      <PersistGate persistor={persistor}>
        <LanguageProvider messages={messages}>
          <ConnectedRouter history={history}>
            <App />
          </ConnectedRouter>
        </LanguageProvider>
      </PersistGate>
    </Provider>,
    MOUNT_NODE
  );
};`

**Comment by shekhardesigner** [comment]
@mouthzipper I followed exact steps above, here is the problem I ran into:

![image](https://user-images.githubusercontent.com/3298585/36030169-403836c2-0dce-11e8-9da8-ff4e3f6c4839.png)

**Comment by mouthzipper** [comment]
@shekhardesigner I forgot to mention that I'm using v4.10.2,
I think my solution won't work on the new version anymore. Need to check the docs in migrating to v5.X

**Comment by billgo** [comment]
@shekhardesigner How do you solve this problem?

**Comment by shekhardesigner** [comment]
@billgo Haven't tried this yet!

**Comment by shekhardesigner** [comment]
@mouthzipper You sure about the version you were using? 4.10.2 doesn't have `session` lib or `PersistGate`

**Comment by mouthzipper** [comment]
@shekhardesigner you are correct, if you decided to use 4.10.2
-- add/edit below codes in configureStore.js
`import { persistStore, autoRehydrate } from 'redux-persist';
...
  const enhancers = [
    applyMiddleware(...middlewares),
    autoRehydrate(),
  ];
....
  persistStore(store);
 return store;
`

**Comment by shekhardesigner** [comment]
I have been trying to integrate V5 of this `redux-persist 5.9.1` with `react boilerplate 3.4.0`. 

Here is what I have done so far:

## 1. Install NPM
`npm i -S redux-persist redux-persist-transform-immutable`

## 2.  Setup Redux Persist in `store.js`

```
//store.js
import .... (other usual stuff)
import { persistStore, persistReducer  } from 'redux-persist';
import storageSession from 'redux-persist/lib/storage/session';
import immutableTransform from 'redux-persist-transform-immutable';

const persistConfig = {
  transforms: [immutableTransform()],
  key: 'root',
  storage: storageSession,
}

const rootReducers = createReducer();

// Using persistReducer not persistCombineReducer because the rootReducer is already returned by combinedReducer from redux-immutable.
const persistedReducer = persistReducer (persistConfig, rootReducers)

export default function configureStore (initialState = {}, history) {
   // other usual stuffs ... 
   
   // I modified how store is created using persistedReducer

   const store = createStore(
      persistedReducer, // this line used to use createReducer() method
      fromJS(initialState),
      composeEnhancers(...enhancers),
   );

   const persistor = persistStore(store);

   return { persistor, store };

   // Please note, I have commented out hot reloading of reducers for now.
}

```

## 3. No change in reducers.js

## 4. Update App.js

```
import 'babel-polyfill';
import React from 'react';

// Added below
import { PersistGate } from 'redux-persist/es/integration/react';

// other usual setup

// Line below used to define just store but now we are defining persistor and store
const { persistor, store } = configureStore(initialState, browserHistory);

// Finally, update the render method:

const render = () => {
  ReactDOM.render(
    <Provider store={store}>
      <PersistGate persistor={persistor}>
        <Router
          history={history}
          routes={rootRoute}
          render={
            applyRouterMiddleware(useScroll())
          }
        />
      </PersistGate>
    </Provider>,
    document.getElementById('app')
  );
};

```

And still no luck:

## Error:

![image](https://user-images.githubusercontent.com/3298585/36751090-4d1de460-1c27-11e8-8834-6818467f9b66.png)

I think I do not have immutable maps configured right. Any help?

**Comment by JonathanLehner** [comment]
:(

**Comment by tilakz** [comment]
@shekhardesigner  have you find any solution? I am having the same issue, please help if you have solved this issue. Thanks in advance.

**Comment by JonathanLehner** [comment]
I mean here are literally 234 open issues.. https://egghead.io/lessons/javascript-redux-persisting-the-state-to-the-local-storage it takes like 5 minutes to setup. Redux persist is a nightmare of complexity and lacks documentation, so if it is not really crucial to save the whole state you might save a lot of time with subscribe.

**Comment by shekhardesigner** [comment]
I agree with @JonathanLehner here, I too didn't implement `redux-persist`. Classic way of `sessionStorage` helped me do a custom implementation

**Comment by tilakz** [comment]
@shekhardesigner @JonathanLehner Thanks guys for your help, the example of Dan Abramov worked out for me.

**Comment by panchicore** [comment]
localStorage.js
```
export const loadState = () => {
  try {
    const serializedState = localStorage.getItem('state')
    if (serializedState === null) {
      return undefined
    }
    return JSON.parse(serializedState)
  } catch (e) {
    console.log('loadStorage error = ', e)
    return undefined
  }
}

export const saveState = (state) => {
  try {
    const serializedState = JSON.stringify(state)
    localStorage.setItem('state', serializedState)
  } catch (e) {
    console.log('saveState error = ', e)
  }
}

```

app.js

```
import {throttle} from 'lodash'
import {loadState, saveState} from './localStorage'

// Create redux store with history
const initialState = loadState()
const store = configureStore(initialState, history)
store.subscribe(throttle(() => {
  //saveState(store.getState()) store the complete state, but I just need the user object so:
  saveState({
    user: store.getState().get('user') // note I am using immutablejs
  })

}, 1000)) // don't persist on disk too often, once per second maybe...
```

works for me, no need of `redux-persist`

**Comment by ygr1k** [comment]
Maybe these links would be helpful:
Issue: https://github.com/rt2zz/redux-persist/issues/1028
PR: https://github.com/rt2zz/redux-persist/pull/1047
